### PR TITLE
More security headers

### PR DESCRIPTION
**Pull request by sj26**
@lox noticed we're missing a couple more headers from https://securityheaders.io/?q=https://buildkite.com — so let's add them!

### CODE REVIEW

**Comment by toolmantim** [comment]
I've rebased this to the latest master, and added the CSP header fix for the screencast videos

**Comment by toolmantim** [comment]
The CSP header is still incomplete, and not read to merge yet. For example, the GA image pings are missing. And there’s still some errors around base64 encoded inline SVGs.

In general, I’d usually say we should go with the nonce approach rather than domain whitelisting… but that seems tricky for a cacheable static site.

**Comment by toolmantim** [comment]
Okay, I've just some updates to finalise the CSP header. I think we should be good to go now 👍🏼

**Comment by toolmantim** [comment]
#100 adds some tests, which gives us a basis for making sure the CSP rules don’t break anything. We might need to add a tear case for the video page though.

File: server.js [hunk 1]
```diff
@@ -24,6 +24,7 @@ const nextHandler = (req, res) => {
 }
 
 const PORT = process.env.PORT || 3000
+const HSTS = process.env.HSTS !== 'false'
```
Comment (sj26) [inline server.js]: @toolmantim This is going to send HSTS headers in dev? I don't think we want that, it could break our dev setups.

**Comment by sj26** [comment]
@toolmantim thanks for the CSP updates!

File: server.js (hunk 1 above)
Comment (toolmantim) [reply server.js]: Ah, good catch! Yeah, I thought I would have needed this for #100, but I don't.

**Comment by toolmantim** [comment]
Updated again!

**Comment by ticky** [comment]
I don’t think it’s possible for us to nonce google analytics, though, given they’ll change it at-will. For what it’s worth, the headers in that page aren’t the same as the headers at https://securityheaders.io/?q=https%3A%2F%2Fbuildkite.com, which is probably what we should be focused on

**Comment by sj26** [comment]
Can we simplify this to just use default-src?

```
default-src 'self' 'unsafe-inline' https://www.google-analytics.com https://d3lj8s78qytm30.cloudfront.net;
```

@ticky sorry I'm not sure I understood this:

> For what it’s worth, the headers in that page aren’t the same as the headers at https://securityheaders.io/?q=https%3A%2F%2Fbuildkite.com, which is probably what we should be focused on

The link you've provided scans the URL we're trying to improve with these headers — the root route is served by site, yeah?

This is served by rails:

https://securityheaders.io/?q=https%3A%2F%2Fbuildkite.com%2Flogin

For rails we do have a CSP but in report only, and we do have referrer policy but only as a `<meta http-equiv>` 😅

**Comment by ticky** [comment]
Sorry, the difference was the link you provided was scanning the headers of _the https redirect_, not the site itself :)

**Comment by sj26** [comment]
Oh, whoops! 🤦🏻‍♂️

**Comment by toolmantim** [comment]
@ticky Hrmm good q. It’s possible, you just nonce the script tag and use `strict-dynamic`. Or are you thinking about sub-resource integrity / digest? For background, in the back-end we follow https://csp.withgoogle.com/docs/strict-csp.html instead of whitelisting, which uses `strict-dynamic`—tho Google Analytics doesn’t need it, it just uses an image tag to do its pings, it doesn’t load any other JS file dynamically. But I don’t understand how nonce’s are meant to work for static/cacheable sites, so we’re using the whitelist approach here. Does that all make sense?

Which headers in which page do you mean though? I thought this PR adds the last two recommended by securityheaders.io? Or did we miss something?

**Comment by toolmantim** [comment]
Ah! Looks like the link in the PR description’s already been updated?

**Comment by sj26** [comment]
This PR was just about adding the two missing headers recommended by securityheaders.io. And imho we shouldn't aim for super locked down, just good enough for now? If we want to do more then I reckon it's another PR, later.

**Comment by sj26** [comment]
@toolmantim I just updated the link in the PR description :-)

**Comment by ticky** [comment]
> Or are you thinking about sub-resource integrity / digest?

Sorry, yeah, my brain was off in SRI land. I suspect we’re okay!

**Comment by toolmantim** [comment]
Woot, okay dokie! Should we merge town?

**Comment by sj26** [comment]
@toolmantim if I stick cloudfront in front of buildkite.com can we move the screencasts to buildkite.com as well? It'd be nice to be using one domain for http2 goodness 💖

**Comment by sj26** [comment]
p.s. https://www2.buildkite.com is currently a cloudfront distribution pointing to the buildkite.com alb, plus https://www2.buildkite.com/assets is pointing at the buildkiteassets.com s3 bucket. I wanted to get it loading same-domain assets when serving www2 to see if it's an improvement, but if we want to go cloudfront anyway maybe we should just flip it first?

**Comment by toolmantim** [comment]
@sj26 yeah for sure! It was only there because the files are so large that the DNS/socket/latency performance penalty was worth the CDN bandwidth wins. Same domain/socket would be even better!

**Comment by sj26** [comment]
Yeah, if we're CDN-ing on the same domain it would be okay I think. Cool!

**Comment by toolmantim** [comment]
Should we set up an ALB rule to point to the screencasts bucket/CloudFront in that case? Otherwise we’d massivley bloat the container image size if we wanted to bake them in, which slows deployments, CI, etc. Hrmm.

**Comment by toolmantim** [comment]
Yeah I reckon flip it!

**Comment by sj26** [comment]
@toolmantim we can add a cloudfront distribution rule to point at the screencasts, and set long expiries, and push them to the edge, independent of the buildkite backend. So best of all worlds! :+1:

**Comment by toolmantim** [comment]
👏🏼

**Comment by toolmantim** [comment]
Oh no. Yeah, we should be able to just add `data:` to the `img-src` rules.

**Pull request merged**